### PR TITLE
Fix Issue 22202 - Wrong error message for implicit call to @system copy constructor in @safe code

### PR DESCRIPTION
**Pull request by RazvanN7**

### CODE REVIEW

File: src/dmd/mtype.d [hunk 1]
```diff
@@ -4743,14 +4743,40 @@ extern (C++) final class TypeFunction : TypeNext
...
-                                m = MATCH.nomatch;
                                 if (pMessage)
                                 {
+                                    /* https://issues.dlang.org/show_bug.cgi?id=22202
+                                     *
+                                     * If a function was deduced by semantic on the CallExp,
+                                     * it means that resolveFuncCall completed succesfully.
+                                     * Therefore, there exists a callable copy constructor,
+                                     * however, it cannot be called because scope constraints
+                                     * such as purity, safety or nogc.
```
Comment (dkorpel) [inline src/dmd/mtype.d:4754]: I would mention that `nothrow` is absent because copy constructors can't throw

**Review by dkorpel — APPROVED** [review]

File: src/dmd/mtype.d (hunk 1 above)
Comment (RazvanN7) [reply src/dmd/mtype.d:4754]: Copy constructors can throw, but nothrow is treated differently and does not suffer from this bug.

Comment (dkorpel) [reply src/dmd/mtype.d:4754]: Oh, then I misremembered

**Review by thewilsonator — APPROVED** [review]

**Comment by ibuclaw** [comment]
This PR introduced a regression https://issues.dlang.org/show_bug.cgi?id=23097

File: src/dmd/mtype.d [hunk 2]
```diff
@@ -4743,14 +4743,40 @@ extern (C++) final class TypeFunction : TypeNext
...
+                                    if (auto f = callExp.f)
+                                    {
+                                        char[] s;
+                                        if (!f.isPure && sc.func.setImpure())
+                                            s ~= "pure ";
+                                        if (!f.isSafe() && !f.isTrusted() && sc.func.setUnsafe())
+                                            s ~= "@safe ";
+                                        if (!f.isNogc && sc.func.setGC())
+                                            s ~= "nogc ";
+                                        s[$-1] = '\0';
```
Comment (kinke) [inline src/dmd/mtype.d:4767]: Probably this line, with an empty `s`. Which would look very bad in the error msg anyway...

**Comment by pbackus** [comment]
This PR introduced a regression: https://issues.dlang.org/show_bug.cgi?id=24301

**Pull request merged**